action_sheet: Add "Mark topic as read" button

lakshya1goel opened the pull request:
This PR adds a "Mark Topic as Read" button to the topic action sheet. The button appears only when there are unread messages in the topic.
  
Fixes: #1225
Fixes: #1357

## Screenshot
![WhatsApp Image 2025-02-06 at 3 05 22 AM](https://github.com/user-attachments/assets/aca7e9bb-2e62-45b0-a093-9cc8be92115b)



### Screen Recording

https://github.com/user-attachments/assets/aaa03f37-a162-4aef-93ba-7115541a5d6a


### Additional
- Updated icon for "Mark all messages as read button" from `Icons.playlist_add_check` to `Icons.mark_chat_read_outlined`

See [CZO Discussion](https://chat.zulip.org/#narrow/channel/530-mobile-design/topic/Icon.20for.20.22Mark.20topic.20as.20read.22.20in.20topic.20action.20sheet)

<table>
  <tr>
    <th>Before</th>
    <th>After</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/e986c386-380e-41a3-9236-fb4fd494d899" width="300"></td>
    <td><img src="https://github.com/user-attachments/assets/974afeec-3e68-4eb6-b350-8c24c627c349" width="300"></td>
  </tr>
</table>

lakshya1goel commented:
Hello, I want to know that should I commit the change of icon of  "Mark all messages as read button" separately?

PIG208 commented:
Yeah, committing it separately sounds good.

PIG208 commented on lib/widgets/action_sheet.dart:
```diff
@@ -372,6 +381,61 @@ class UserTopicUpdateButton extends ActionSheetMenuItemButton {
+    final zulipLocalizations = ZulipLocalizations.of(pageContext);
+
+    try {
+      if (connection.zulipFeatureLevel! >= 155) {
```
`markNarrowAsRead` handles feature level compatibility and only takes a `BuildContext` and a `Narrow`. We can use that here.

lakshya1goel replied:
Thanks for the review, I have updated the implementation and thus, removed the tests as well targeting the API requests.

PIG208 reviewed (commented):
Thanks for the PR! Took a quick look at the implementation and left a comment.

alya commented:
Why is the icon different from the one we have in the web app?

![Screenshot 2025-01-31 at 11 01 39](https://github.com/user-attachments/assets/0ba470e9-066c-494b-af5b-2dc702bae111)

lakshya1goel commented:
Thanks Alya for pointing. I have updated the correct icon and also updated the PR description with latest screenshots.

lakshya1goel commented:
Thanks for the review @PIG208, pushed the revision, PTAL.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
     });
 
+    group('MarkTopicAsReadButton', () {
+      Future<void> setupToTopicActionSheetWithUnreadMessages(WidgetTester tester, {
```
Looks like this and `setupToTopicActionSheetWithNoUnreadMessages` are each only called once in their respective test. Let's reorganize this for clarity.

The logic tested here whether `MarkTopicAsReadButton` is hidden in the action sheet, where the only variable is the presence of unread messages in the store, so it would make sense to have a single helper that's capable of controlling that.

If you look at the "topic action sheet" test group, you can find a helper called `prepare` that should accomplish all of these for you. You can use the "UserTopicUpdateButton" test group as a reference.

PIG208 replied:
The following comments apply to this helper, but you can skip most of them when working on the next revision.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
+      }) async {
+        addTearDown(testBinding.reset);
+
+        final effectiveChannel = channel ?? eg.stream();
```
Because `channel` is unused, we can just have `final channel = eg.stream()` here.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
 
+    group('MarkTopicAsReadButton', () {
+      Future<void> setupToTopicActionSheetWithUnreadMessages(WidgetTester tester, {
+        ZulipStream? channel,
```
We can remove this because it is unused.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
+        await store.addMessage(message);
+        store.unreads.streams[effectiveChannel.streamId] ??= {};
+        store.unreads.streams[effectiveChannel.streamId]![topicName] ??= QueueList<int>();
+        store.unreads.streams[effectiveChannel.streamId]![topicName]!.add(message.id);
```
We shouldn't need to modify `store.unreads.streams` here. A convenient way to add unread messages to the store is through `store.handleEvent(MessageEvent(id: 0, message: message))`. There should be some examples in other tests.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
+        connection = store.connection as FakeApiConnection;
+
+        connection.prepare(json: eg.newestGetMessagesResult(
+          foundOldest: true, messages: [message]).toJson());
```
Looking at `setupToMessageActionSheet`, the `connection.prepare` call is right before `tester.pumpWidget`. This is because this call prepares for a request that will be made by `MessageListPage`. Moving it there help highlight their relationship.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
+        await tester.pump(const Duration(milliseconds: 250));
+      }
+
+      Future<void> setupToTopicActionSheetWithNoUnreadMessages(WidgetTester tester) async {
```
The other helper can be removed once we refactor the first one to handle both situations.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
+          await setupToTopicActionSheetWithUnreadMessages(tester);
+
+          final zulipLocalizations = GlobalLocalizations.zulipLocalizations;
+          check(find.text(zulipLocalizations.actionSheetOptionMarkTopicAsRead)).findsOne();
```
Let's use the string directly, without referencing `GlobalLocalizations`, because we don't need tests to respond to locale changes and that we can be clear about what is expected exactly.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -1061,6 +1061,82 @@ void main() {
       });
     });
 
+    group('MarkTopicAsReadButton', () {
```
This test should be moved to the "topic action sheet" group.

PIG208 reviewed (commented):
Thanks for the PR! Went through the changes and left some comments, mainly on the tests.

lakshya1goel commented:
Thanks for the review @PIG208, I have updated the tests and the commit message accordingly. Please have a look and let me know if anything else is required.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -450,6 +450,39 @@ void main() {
+      Future<void> setupToTopicActionSheet(WidgetTester tester, {
+        required bool hasUnreadMessages,
+      }) async {
+        addTearDown(testBinding.reset);
```
We can skip the `addTearDown` call because `prepare` already calls it for us.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -450,6 +450,39 @@ void main() {
+          channel: channel, topic: topic, messages: [message]);
+      }
+
+      testWidgets('visible if topic has unread messages', (tester) async {
+        await setupToTopicActionSheet(tester, hasUnreadMessages: true);
+        check(find.text('Mark Topic As Read')).findsOne();
+      });
```
Let's try to simplify away the helper, we can always add it back if we find a lot of tests reusing the same logic:

```suggestion
      testWidgets('visible if topic has unread messages', (tester) async {
        await prepare();
        final message = eg.streamMessage(stream: someChannel, topic: someTopic);
        await store.handleEvent(MessageEvent(id: 0, message: message));
        await showFromAppBar(tester, messages: [message]);
        check(find.text('Mark Topic As Read')).findsOne();
      });
```

This way we can also skip the flag and the if-statement.

PIG208 reviewed (commented):
Thanks for the update! Left a review comment.

lakshya1goel commented:
Pushed the changes, Please have a look now @PIG208 , Thanks!

lakshya1goel commented on test/widgets/action_sheet_test.dart:
```diff
@@ -450,6 +450,39 @@ void main() {
+        await setupToTopicActionSheet(tester, hasUnreadMessages: false);
+        check(find.text('Mark Topic As Read')).findsNothing();
+      });
+    });
```
Should I add 2 tests for both:  
- For FL < 155: /api/v1/mark_topic_as_read
- For FL ≥ 155: /api/v1/messages/flags/narrow

PIG208 replied:
Here we don't need to, because the tests for `markNarrowAsRead` cover that.

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -450,6 +450,53 @@ void main() {
+            'op': 'add',
+            'flag': 'read',
+          });
+      });
```
For this new test, we just need to focus on checking the field most interesting to us, i.e. 'narrow'. This is exactly what we are doing in 'on topic move, acts on new topic':

```dart
        testWidgets('on topic move, acts on new topic', (tester) async {
          // ...
          check(connection.lastRequest).isA<http.Request>()
            ..method.equals('POST')
            ..url.path.equals('/api/v1/messages/flags/narrow')
            ..bodyFields['narrow'].equals(
                jsonEncode(eg.topicNarrow(newStream.streamId, newTopic).apiEncode()));
```

This is helpful to highlight what is important for this particular test.

lakshya1goel replied:
Thanks for pointing, I will make sure to include the necessary part only from now on.

gnprice replied:
This needs more checks than just the narrow — see https://github.com/zulip/zulip-flutter/pull/1274#discussion_r1960914494.

One key point about that "on topic move, acts on new topic" test is that it comes after a smoke-test for the same button, which checks all the details including that the request is marking the messages in question as (in that case) unread. So that test is focused specifically on the part that's different when the topic moved, and that difference is all about the narrow that's being acted on.

PIG208 commented:
Could you also post a screenshot of the topic action sheet with the mark topic as read button on there? For design changes, we prefer screenshots over screen recording, unless it is specific about some interaction. Even then, having both included is helpful.

lakshya1goel commented:
Updated the description of PR.

alya commented:
The capitalization doesn't match the web app / issue. Please be mindful not to introduce unexpected changes from the spec.

lakshya1goel commented:
> The capitalization doesn't match the web app / issue. Please be mindful not to introduce unexpected changes from the spec.

Sorry for missing out that, updated it. Also changed the screenshot and recording in PR description.

chrisbobbe commented on lib/widgets/action_sheet.dart:
```diff
@@ -372,6 +380,30 @@ class UserTopicUpdateButton extends ActionSheetMenuItemButton {
+    return zulipLocalizations.actionSheetOptionMarkTopicAsRead;
+  }
+
+  @override void onPressed() async {
+    if (!pageContext.mounted) return;
+    await markNarrowAsRead(pageContext, TopicNarrow(channelId, topic));
+  }
```
This should have error handling, like other buttons that make API requests

lakshya1goel replied:
Since the `markNarrowAsRead` itself has its error handling, I haven't handled that here. Following the [CZO discussion](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/.23F1317.20showErrorDialog), I have pushed the revision.

PIG208 commented:
There are some conflicts since the previous update. Could you resolve that? Thanks!

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -568,6 +568,44 @@ void main() {
+        check(connection.lastRequest).isA<http.Request>()
+          ..url.path.equals('/api/v1/messages/flags/narrow')
+          ..bodyFields['narrow'].equals(jsonEncode([
+            ...TopicNarrow(someChannel.streamId, TopicName(someTopic)).apiEncode(),
+            {'operator': 'is', 'operand': 'unread'}
+          ]));
```
nit: indentation
```suggestion
              ...TopicNarrow(someChannel.streamId, TopicName(someTopic)).apiEncode(),
              {'operator': 'is', 'operand': 'unread'}
            ]));
```

PIG208 reviewed (commented):
Thanks for the update! This mostly looks good to me, with some nits.

chrisbobbe commented on lib/widgets/message_list.dart:
```diff
@@ -852,7 +852,7 @@ class _MarkAsReadWidgetState extends State<MarkAsReadWidget> {
               ),
               onPressed: _loading ? null : () => _handlePress(context),
-              icon: const Icon(Icons.playlist_add_check),
+              icon: const Icon(ZulipIcons.message_checked),
```
Ah: I missed the relevant discussion and ended up opening an issue for this change 🙂 please add `Fixes #1357.` to the commit message and PR description. Thanks!

lakshya1goel commented:
Pushed the revision @PIG208, PTAL. Thanks!

PIG208 commented on test/widgets/action_sheet_test.dart:
```diff
@@ -568,6 +568,44 @@ void main() {
+          ..url.path.equals('/api/v1/messages/flags/narrow')
+          ..bodyFields['narrow'].equals(jsonEncode([
+              ...TopicNarrow(someChannel.streamId, TopicName(someTopic)).apiEncode(),
+              {'operator': 'is', 'operand': 'unread'}
```
nit: trailing comma for list elements

PIG208 commented:
Looks good except for one more nit! Marking it for Greg's review.

lakshya1goel commented:
Hi @gnprice, please review the PR when you have some time, thanks!

chrisbobbe commented on lib/widgets/actions.dart:
```diff
@@ -18,49 +18,59 @@ import '../model/narrow.dart';
+///
+/// Methods in this class show progress and error feedback in the UI,
+/// and handle error cases internally rather than expecting the caller to handle them.
+abstract final class ZulipAction {
```
`markNarrowAsUnreadFromMessage` should also go in this class; see PR #1317 for a revised version of this commit.

gnprice commented on test/widgets/action_sheet_test.dart:
```diff
@@ -568,6 +568,44 @@ void main() {
+        await tester.tap(find.text('Mark topic as read'));
+        await tester.pumpAndSettle();
+
+        check(connection.lastRequest).isA<http.Request>()
+          ..url.path.equals('/api/v1/messages/flags/narrow')
+          ..bodyFields['narrow'].equals(jsonEncode([
+              ...TopicNarrow(someChannel.streamId, TopicName(someTopic)).apiEncode(),
+              {'operator': 'is', 'operand': 'unread'},
+            ]));
```
This checks it's doing *something* to the message flags in this narrow — but not that it's marking as read. In particular this test would equally well pass if the implementation were changed to mark the messages as *unread* instead of read.

Let's add the checks that ensure what it's doing is marking as read, rather than marking as unread or setting or unsetting some other flag.

gnprice commented on lib/widgets/actions.dart:
```diff
@@ -18,49 +18,59 @@ import '../model/narrow.dart';
+  /// Marks all messages in the given [narrow] as read, showing feedback in the UI.
+  ///
+  /// Shows progress feedback for long-running operations, and error feedback
+  /// if the operation fails. Updates unread counts in the UI on success.
+  static Future<void> markNarrowAsRead(BuildContext context, Narrow narrow) async {
```
This last sentence isn't accurate — this function doesn't do anything to unread counts in the UI.

(It wouldn't, because it's the model's job to drive the data that appears in the UI. But this doesn't do anything to unread counts in the model, either. If you're thinking of the `PerAccountStoreWidget.of(context).unreads.handleAllMessagesReadSuccess()` call at the end, see that method's doc.)

gnprice commented on lib/widgets/actions.dart:
```diff
@@ -18,49 +18,59 @@ import '../model/narrow.dart';
-        title: zulipLocalizations.errorMarkAsReadFailedTitle,
-        message: e.toString()); // TODO(#741): extract user-facing message better
-      return;
+/// High-level operations that interact with the Zulip API and handle UI feedback.
```
"handle feedback" sounds like something that *receives* feedback and does something about it. These do the opposite, showing feedback to the user.

gnprice reviewed (commented):
Thanks @lakshya1goel for taking care of this, and thanks @PIG208 and all for the previous reviews! Comments below.

Let's leave out the last commit introducing ZulipAction; that change is mostly a matter of writing docs, so I think it'll be simpler if I just do that change myself rather than going through cycles of editing.

gnprice commented:
> Let's leave out the last commit introducing ZulipAction; that change is mostly a matter of writing docs, so I think it'll be simpler if I just do that change myself rather than going through cycles of editing.

I've just sent #1366 for this part.

lakshya1goel commented on lib/widgets/action_sheet.dart:
```diff
@@ -461,6 +469,30 @@ class ResolveUnresolveButton extends ActionSheetMenuItemButton {
+    return zulipLocalizations.actionSheetOptionMarkTopicAsRead;
+  }
+
+  @override void onPressed() async {
+    if (!pageContext.mounted) return;
+    await ZulipAction.markNarrowAsRead(pageContext, TopicNarrow(channelId, topic));
```
No, there was no discussion done for this in previous round. Also I will not mark the comments as resolved form now on.

lakshya1goel commented:
Pushed the revision - PTAL @gnprice , Thanks!

lakshya1goel commented:
Hello @gnprice, whenever you have some time, could you kindly take a look at the PR? Thank you!

gnprice commented on assets/icons/message_checked.svg:
nit: commit message has stray leading space

     msglist: Update "Mark all messages as read" button icon
    
    Icon taken from the Figma design:
    https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=3512-7&t=FEydAZe5DNKVlKnV-0
    
    Fixes: #1357

gnprice commented on test/widgets/action_sheet_test.dart:
```diff
@@ -562,6 +563,47 @@ void main() {
+
+      testWidgets('not visible if topic has no unread messages', (tester) async {
+        await prepare();
+        final message = eg.streamMessage(stream: someChannel, topic: someTopic, flags: [MessageFlag.read]);
```
nit: line too long: key information is past 80 columns

gnprice reviewed (commented):
Thanks for the revision! All looks good, modulo two nits below. I'll fix them up and merge.

gnprice commented on test/widgets/action_sheet_test.dart:
```diff
@@ -562,6 +563,47 @@ void main() {
+
+      testWidgets('marks topic as read when pressed', (tester) async {
+        await prepare();
+        final message = eg.streamMessage(stream: someChannel, topic: someTopic);
+        await store.handleEvent(MessageEvent(id: 0, message: message));
+        await showFromAppBar(tester, messages: [message]);
```
Hmm I guess there's one more-substantive comment — this test should be fixed for some of the same comments made in the last review above.

I've already fixed the two nits (and also rebased past changes I merged earlier today), so I'll push those fixes back to the PR branch. But then @lakshya1goel  please make a revision to update this test case too based on the feedback given above on other tests.

lakshya1goel commented:
Thanks @gnprice for updating the PR. I have also pushed the revision for the test case. PTAL, Thanks!

gnprice commented:
Thanks! Looks good; merging.
